### PR TITLE
Add noise options to ACES

### PR DESCRIPTION
**Pull request by epelaaez**
Adds the ability to pass noise models to the ACES endpoints to allow for more interesting simulations.

(related: https://github.com/Infleqtion/server-superstaq/pull/2634)

### CODE REVIEW

File: general-superstaq/general_superstaq/service.py
Comment (bharat-thotakura) [inline general-superstaq/general_superstaq/service.py:327]: nit
```suggestion
                    `noise = ("asymmetric_depolarize", (0.1, 0.1, 0.1))`. Notice that these values
```
and the two other occurrences

Comment (epelaaez) [reply general-superstaq/general_superstaq/service.py]: done!

File: general-superstaq/general_superstaq/service.py
Comment (richrines1) [inline general-superstaq/general_superstaq/service.py:322]: * imo the noise model and error probability should be two separate arguments, instead of having to be passed as a tuple
* `cirq.NoiseModel`s are json-serializable - would it make sense to allow them to be passed as well?

also - predates this pr, but all the parameters we're documenting here they should really be made into optional arguments instead of using kwargs

Comment (vtomole) [reply general-superstaq/general_superstaq/service.py:322]: > cirq.NoiseModels are json-serializable - would it make sense to allow them to be passed as well?

Makes sense. This will have to be hacky on the qiskit side though.

Comment (epelaaez) [reply general-superstaq/general_superstaq/service.py:322]: > imo the noise model and error probability should be two separate arguments, instead of having to be passed as a tuple

Sounds good.

> cirq.NoiseModels are json-serializable - would it make sense to allow them to be passed as well?

I was thinking about doing this, but I don't think there's an exact equivalence between Qiskit's and Cirq's models, so it isn't as easy to convert back and forth. Or we could allow `cirq.NoiseModel` coming from css and keep it as it is for qss.

> predates this pr, but all the parameters we're documenting here they should really be made into optional arguments instead of using kwargs

I'll change them to optional parameters.

Comment (bharat-thotakura) [reply general-superstaq/general_superstaq/service.py:322]: > Or we could allow cirq.NoiseModel coming from css and keep it as it is for qss.

Can we not use `NoiseModel.to_dict()` and `from_dict()` for qss?

Comment (epelaaez) [reply general-superstaq/general_superstaq/service.py:322]: > Can we not use NoiseModel.to_dict() and from_dict() for qss?

Ah yeah, I think this could work.

Comment (epelaaez) [reply general-superstaq/general_superstaq/service.py:322]: For reference, I think the best way to translate from Qiskit's to Cirq's noise models is to use [`Kraus(qiskit_model)`](https://qiskit.org/documentation/stubs/qiskit.quantum_info.Kraus.html) and then use [`cirq.KrausChannel`](https://quantumai.google/reference/python/cirq/KrausChannel) to go back and forth. Let me know if you think there is a better way to do this.

Comment (richrines1) [reply general-superstaq/general_superstaq/service.py:322]: > (even tag and lifespan? these two are also optional when creating jobs)

imo they should be optional arguments in `create_job` too, if they are intended for external use - if nothing else it's a form of documentation (part of #523)

Comment (epelaaez) [reply general-superstaq/general_superstaq/service.py:322]: > imo they should be optional arguments in create_job too

sounds good! I'll make the change for this function for now.

Comment (richrines1) [reply general-superstaq/general_superstaq/service.py:322]: > use [`cirq.KrausChannel`](https://quantumai.google/reference/python/cirq/KrausChannel) to go back and forth

`qss` shouldn't be given a `cirq` dependency, unfortunately

i kinda think it's okay for parity to be imperfect here though (or if nothing else it seems preferable to keeping a useful functionality out of one or the other client just for the sake of parity). We'll still be able to specify a noise model in `qss` using the string method. I imagine that we'll also be making our error models more general/customizable, which would hopefully expand this to cover a lot more use cases

we can also add an issue to revisit this in the future

Comment (epelaaez) [reply general-superstaq/general_superstaq/service.py:322]: > qss shouldn't be given a cirq dependency, unfortunately

Oh, I meant server-side, sorry I wasn't specific. In the client I think it's possible to let the users send a serialized `cirq.NoiseModel` and a `qiskit_aer.noise.QuantumError`.

Comment (epelaaez) [reply general-superstaq/general_superstaq/service.py:322]: It looks like [`qiskit_aer.noise.NoiseModel.from_dict`](https://qiskit.org/ecosystem/aer/_modules/qiskit_aer/noise/noise_model.html#NoiseModel.from_dict) is being deprecated, and I don't think there's another way to reconstruct a Qiskit noise model back in the server.

**Comment by epelaaez** [comment]
I added support to send a `cirq.NoiseModel` through `css.Service.submit_aces`. I wanted to do the same for `qss.SuperstaqBackend`, but the methods in Qiskit to initialize a noise model from a dictionary (which we need to construct the model back in the server) are being deprecated. For now, I've only left the original noise implementation in qss.

Related, now that the functionality is different between css and qss, should `qss.SuperstaqBackend.submit_aces` move to `qss.SuperstaqProvider`. Since, if I understood correctly, the rationale under putting it in `SuperstaqBackend` was that css and qss did the same (original discussion: https://github.com/Infleqtion/client-superstaq/pull/687#discussion_r1291559327, cc: @richrines1).

**Comment by richrines1** [comment]
> I added support to send a `cirq.NoiseModel` through `css.Service.submit_aces`. I wanted to do the same for `qss.SuperstaqBackend`, but the methods in Qiskit to initialize a noise model from a dictionary (which we need to construct the model back in the server) are being deprecated. For now, I've only left the original noise implementation in qss.

ok sounds good. we'll definitely want to generalize this more in the future anyway (e.g. for standard simulations with noise)
 
> Related, now that the functionality is different between css and qss, should `qss.SuperstaqBackend.submit_aces` move to `qss.SuperstaqProvider`

yep i think so

File: cirq-superstaq/cirq_superstaq/service.py
Comment (epelaaez) [inline cirq-superstaq/cirq_superstaq/service.py:848]: So, apparently not **all** `cirq.NoiseModel`s are serializable (see: https://github.com/quantumlib/Cirq/issues/6249). Should I add a `try` statement here to catch when a given noise model is not serializable? Or is the error thrown by Cirq enough?

Error thrown by Cirq:
```
    160 def default(self, o):
    161     """Implement this method in a subclass such that it returns
    162     a serializable object for ``o``, or calls the base implementation
    163     (to raise a ``TypeError``).
   (...)
    177 
    178     """
--> 179     raise TypeError(f'Object of type {o.__class__.__name__} '
    180                     f'is not JSON serializable')

 

TypeError: Object of type InsertionNoiseModel is not JSON serializable

```

File: qiskit-superstaq/qiskit_superstaq/superstaq_backend.py
Comment (richrines1) [inline qiskit-superstaq/qiskit_superstaq/superstaq_backend.py:426]: why was this removed from `SuperstaqBackend`?

Comment (epelaaez) [reply qiskit-superstaq/qiskit_superstaq/superstaq_backend.py:426]: I thought we wanted to move it from `SuperstaqBackend` to `SuperstaqProvider` (https://github.com/Infleqtion/client-superstaq/pull/699#issuecomment-1679753907).

Comment (richrines1) [reply qiskit-superstaq/qiskit_superstaq/superstaq_backend.py:426]: ohhh im sorry i totally misread that question.

basically i think we want the methods to appear in \*both\* `qss.SuperstaqProvider` and `qss.SuperstaqBackend`. The reason we only had to put them in `qss.SuperstaqBackend` in #687 was that `SuperstaqProvider` was already inheriting them from `gss.Service`

thinking about this a bit more though - it seems like the `qss` version of this is still independent of qiskit? in which case we might not want to remove it from `gss.Service`; instead we can just:
* leave `submit_aces` in `gss.Service` and `qss.SuperstaqBackend`, but update both include the new/updated arguments
* override `submit_aces` in `css.Service` to add support for cirq noise models
* leave `process_aces` alone

**Comment by stephanielee9** [comment]
copy of https://github.com/Infleqtion/client-superstaq/pull/713

**Comment by epelaaez** [comment]
@vtomole @richrines1 This PR now has the changes that were made in the PR form my fork.

**Review by vtomole — APPROVED** [review]
LGTM assuming this has the same contents as https://github.com/Infleqtion/client-superstaq/pull/713

**Comment by epelaaez** [comment]
I'm going to re-check, but I think they do in fact have the same changes. I will also wait for the server-side changes since otherwise the integration test will fail.

File: qiskit-superstaq/qiskit_superstaq/daily_integration_test.py
Comment (vtomole) [inline qiskit-superstaq/qiskit_superstaq/daily_integration_test.py:240]: This throws `FAILED qiskit-superstaq/qiskit_superstaq/daily_integration_test.py::test_aces - TypeError: Object of type QuantumRegister is not JSON serializable` when i run it on the local server. Why is it not 
```suggestion
        qubits=[0],
```

Like the `test_aces` in `cirq-superstaq?

Comment (epelaaez) [reply qiskit-superstaq/qiskit_superstaq/daily_integration_test.py:240]: Yes this should be `qubits=[0]`, I didn't update this test, sorry about that. I changed the `qubits` argument from accepting a `qiskit.QuantumRegister` to a `Sequence[int]` since it is harder to specify non-linear qubits with `qiskit.QuantumRegister`.

**Comment by vtomole** [comment]
@epelaaez The server PR that this corresponds to has been merged so you can merge this in as well.

**Comment by epelaaez** [comment]
I don't think I have write access to merge this in, so feel free to do so @vtomole !

**Pull request merged**
